table/block_based_table_reader.h

```diff
@@ -627,7 +627,10 @@ class BlockBasedTableIterator : public InternalIterator {
   const InternalKeyComparator& icomp_;
   InternalIterator* index_iter_;
   PinnedIteratorsManager* pinned_iters_mgr_;
-  BlockIter data_block_iter_;
+  BlockIter* block_iter_;
+  // Only one of these two will be used. block_iter_ will refer to the used one.
+  DataBlockIter data_block_iter_;
+  IndexBlockIter index_block_iter_;
```

Comment:
By moving the function calls from a final class to base case, the virtual function will be back. I don't like template but in this case, can we do template instead?

Reply:
oh yes but because of is_index case it does not seem like we can use templates. perhaps we can discard the whole PR if you think it would hurt performance because of removing the final flag.

Reply:
Can you explain more why is_index makes template impossible?

Reply:
come to think of it, it is not impossible, the template argument will just be propagates up all the way to BlockBasedTableIterator. i could go for that if that is what you meant.